dask/distributed/worker.py

```diff
@@ -427,16 +432,27 @@ def collect(self, locations):
         qkey = str(uuid.uuid1())
         queue = Queue()
         self.queues[qkey] = queue
+        locations_copy = locations.copy()  # don't mutate global locations
 
         # Send out requests for data
-        log(self.address, 'Collect data from peers', locations)
+        log(self.address, 'Collect data from peers', locations_copy)
         start = time()
         counter = 0
         with logerrors():
-            for key, locs in locations.items():
+            for key, locs in list(locations_copy.items()):
                 if key in self.data:  # already have this locally
+                    locations_copy.pop(key)
                     continue
                 worker = random.choice(tuple(locs))  # randomly select one peer
+
+                # track keys and where they are comming from
+                if worker not in self.queues_by_worker:
+                    self.queues_by_worker.update({worker: {qkey: [key]}})
+                elif qkey not in self.queues_by_worker[worker]:
+                    self.queues_by_worker[worker].update({qkey: [key]})
+                elif key not in self.queues_by_worker[worker][qkey]:
+                    self.queues_by_worker[worker][qkey].append(key)
```

Comment:
taking a look at this now.  A couple of alternatives might be

1.  use a `defaultdict`
2.  Use if statements to set up the right context, then add something at the end, e.g.

```python
if worker not in self.queues_by_worker:  # alternatively use defaultdict
    self.queues_by_worker[worker] = dict()
if qkey not in self.queues_by_worker[worker]:
    self.queues_by_worker[worker][qkey] = set()
self.queues_by_worker[worker][qkey].add(key)
```

Or define `queues_by_worker` cleverly as something like the following:

    self.queues_by_worker = defaultdict(lambda : defaultdict(lambda: defaultdict(set)))

And then just state the following in all cases without checking

    self.queues_by_worker[worker][qkey].add(key)

Though this might be more confusing.  I didn't check the above.  It is likely incorrect.